### PR TITLE
KSP: Allow to specify extra all-open annotations

### PR DESCRIPTION
**Pull request by dstepanov**
Workaround for https://github.com/micronaut-projects/micronaut-core/issues/9764

### CODE REVIEW

**Review by graemerocher — APPROVED** [review]

**Review by yawkat — APPROVED** [review]

File: inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/AbstractKotlinElement.kt [hunk 1]
```diff
@@ -605,6 +610,8 @@ internal abstract class AbstractKotlinElement<T : KotlinNativeElement>(
     }
 
     companion object {
+        val extraOpenAnnotations = System.getProperty("kotlin.allopen.annotations")?.split(",")?.toTypedArray()
```
Comment (Spikhalskiy) [inline inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/AbstractKotlinElement.kt:613]: Can we maybe get in as a part of the gradle plugin configuration too?

Comment (dstepanov) [reply inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/AbstractKotlinElement.kt:613]: Probably yes, maybe even have it automatically if the allopen plugin exists @melix, right?
Let's first see that it's working.

**Review by sdelamo — CHANGES_REQUESTED** [review]
Please, document the system property`kotlin.allopen.annotations`

**Comment by dstepanov** [comment]
@sdelamo Let's first check that it is working, and then I can document it. We might need to add it for KAPT, too, if the next version is broken.

**Comment by sdelamo** [comment]
> @sdelamo Let's first check that it is working, and then I can document it. We might need to add it for KAPT, too, if the next version is broken.

Please document it in this pull request. I don't want to have secret system properties in the code. I assume you have tested it, and it works. If, for whatever reason, we have to roll it back, then we can roll back the squashed commit resulting from this PR, including documentation.

**Comment by dstepanov** [comment]
@sdelamo Added

**Review by sdelamo — APPROVED** [review]

**Comment by Spikhalskiy** [comment]
@dstepanov It looks like it's not easy to make an environment variable with dots in it. Can we change it to screaming case?

**Comment by dstepanov** [comment]
You can do it this way like in the OpenApi:
```
tasks.withType(JavaCompile) {
    options.fork = true
    options.forkOptions.jvmArgs << '-Dmicronaut.openapi.property.naming.strategy=SNAKE_CASE'

    ...
}
```

**Comment by dstepanov** [comment]
https://micronaut-projects.github.io/micronaut-openapi/latest/guide/#systemPropertyConfiguration

**Comment by Spikhalskiy** [comment]
@dstepanov OpenApi uses https://github.com/micronaut-projects/micronaut-openapi/blob/adc7d7c9554aa1fab8e79657d6a4a60ec92339c4/openapi/src/main/java/io/micronaut/openapi/visitor/ConfigUtils.java#L732 to read the property. I don't think System.getProperty will return you the property supplied with -D.

My results:
1. Snapshot micronaut, kotlin2.0-Beta2 with kapt and kapt.use.k2=true, supplying "-Dkotlin.allopen.annotations=jakarta.inject.Singleton" has no effect, it throws "Cannot apply AOP advice to final class. Class must be made non-final to support proxying" 
2. Snapshot micronaut, Kotlin 1.9.22 with ksp hangs for me on kspKotlin with the latest micronaut snapshot. With or without the property. It doesn't throw an error it used to throw about the final class after 30 minutes.

**Comment by dstepanov** [comment]
It's only supported for KSP for now

**Comment by Spikhalskiy** [comment]
I posted CPU sampling results of my ksp hanging in the KAPT -> KSP performance degradation issue: https://github.com/micronaut-projects/micronaut-core/issues/9750#issuecomment-1892652585
It looks like there is a severe ksp performance degradation between 4.2.x and 4.3.0-SNAPSHOT.

**Comment by Spikhalskiy** [comment]
Update, I gave kspKotlin 40 minutes and it failed with the same error as before
```
Method defines AOP advice but is declared final. Change the method to be non-final in order for AOP advice to be applied.
```

The following is in gradle.properties:
```
kotlin.daemon.jvmargs=-Xmx4096M -Dkotlin.allopen.annotations="jakarta.inject.Singleton"
org.gradle.jvmargs=-Xmx4096M -Dkotlin.allopen.annotations="jakarta.inject.Singleton"
```

**Pull request merged**